packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,149 @@
+  fetchFunction: (params: ServerSideParams) => Promise<{
+    agentsList: Array<Agent>;
+    totalCount: number;
+    startKey: string;
+  }>
+) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: getDefaultSort(),
+  });
+  const [pageCount, setPageCount] = useState({
+    to: 0,
+    from: 0,
+    total: 0,
+  });
+
+  const limit = 15;
+
+  // startKey is used here as a way to paginate through agents returned from
+  // their respective rpcs.
+  const [fetchAttempt, fetch] = useAsync(async (startKey: string) =>
+    retryWithRelogin(ctx, documentUri, clusterUri, () =>
+      fetchFunction({
+        ...agentFilter,
+        limit,
+        clusterUri,
+        startKey,
+      })
+    )
+  );
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        fetch(proposedKey);
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        // if we got pageIndex - 1, we'd get the same data as the previous "next" we just called.
+        // we have to go back 2. which is why we use newPageIndex instead
+
+        // example after going from pageIndex = 0 to pageIndex = 1:
+        // ['startKeyFromInitialFetch', 'goEvenFurtherStartKey']
+        //     ^- just came from here          ^- we are here
+        //
+        // so in order to get the start key for the page BEFORE this one, we go back twice.
+        // this could also be written/read as keys[pageIndex - 2]
+        fetch(keys[newPageIndex - 1]);
+        setPageIndex(newPageIndex);
+      };
+    }
+    return null;
+  }
+
+  useEffect(() => {
+    // when agent filter changes we consider it a 'new' search so we should pass an empty
+    // startKey to get the beginning of the list with new filters, and set pageIndex back to zero
+    fetch('');
+    setPageIndex(0);
+  }, [agentFilter]);
+
+  // when we receive data from fetch, we store the startKey (or lack of) according to the current
+  // page index. think of this as "this page's nextKey".
+  // "why don't we just name it nextKey then?"
+  // Mostly because it's called startKey almost everywhere else in the UI, and also because we'd have the same issue
+  // for prevPage if we swapped named, and this comment would be explaining "this page's startKey".
+  useEffect(() => {
+    const newKeys = [...keys];
+    newKeys[pageIndex] = fetchAttempt.data?.startKey;
+    setKeys(newKeys);
+
+    setPageCount(getPageCount());
+  }, [fetchAttempt.data]);
```

Comment:
This falls under chains of computations that I mentioned above.

https://beta.reactjs.org/learn/you-might-not-need-an-effect#chains-of-computations

Reply:
> These two pieces of data can only be updated after the data has been received as that returned data contains the information from the server regarding page metadata and keys.

Right, so we not update them when you receive that data?

The docs say:

> Sometimes you might feel tempted to chain Effects that each adjust a piece of state based on other state.

And I think this is what we do here: we wait for `fetchAttempt` to get updated so that we can adjust keys but we could do it after we receive a response from `fetch`.

In that effect where you execute `fetch`, you could do something like

```typescript
useEffect(() => {
  const fetchAndUpdateKeys = async () => {
    const [response, err] = await fetch(keys[pageIndex - 1]);

    if (err) {
      return;
    }

    const startKey = response.startKey;

    // Do something with keys.
  };

  fetchAndUpdateKeys();
}, [agentFilter, pageIndex]);
```

It'd be nice to pass `agentFilter` directly to `fetch` too so that the dependency between them is clearly visible when one calls `fetch`.

Reply:
But in general I have to say this hook looks better and better with every change! The current version is much, much less intimidating than the initial one.

Reply:
I see the forest through the trees now and understand what you mean. Great change and even easier to read